## Association Rule Mining in Retail Store

### Problem Statement:
 * What are the items that may be frequently purchased together?

### Objective:
* To learn how Apriori Algorithm and Association Rules works.
* To learn how Combination and Permutation helps to find Support and Confidence of itemsets respectively.
* To find frequent itemsets with high confidence and lift, keeping both item together will help to increase sales.


### Introduction
* Association rule mining is one of an important technique of data mining for knowledge discovery.
* The knowledge of the correlation between the items in the data transaction can use association rule mining.
* Retail store analysis is one of an application area of association rule mining technique.
* The possible percentage of the correlation of combined items gives the new knowledge. Therefore, it is a very helpful for determiner to take the decisions

### Analysis

In [24]:
## Importing Required Library

import pandas as pd
import numpy as np

In [4]:
# Reading Excel file 
bread = pd.read_excel('raw_bread.xlsx')

In [5]:
## Here we have transaction data, which include column, Date,Time,Transaction,Item
## we should remove duplicate transaction, it shows quantity of item in same transaction,
# it is not needed in appriori aglo as we only care about different item in particular transaction
bread

Date,Time,Transaction,Item
0              2016-10-30,09:58:11,1,Bread
1       2016-10-30,10:05:34,2,Scandinavian
2       2016-10-30,10:05:34,2,Scandinavian
3      2016-10-30,10:07:57,3,Hot chocolate
4                2016-10-30,10:07:57,3,Jam
...                                    ...
21288      2017-04-09,14:32:58,9682,Coffee
21289         2017-04-09,14:32:58,9682,Tea
21290      2017-04-09,14:57:06,9683,Coffee
21291      2017-04-09,14:57:06,9683,Pastry
21292   2017-04-09,15:04:24,9684,Smoothies

[21293 rows x 1 columns]

In [6]:
## dropping Duplicate Transaction
bread = bread.drop_duplicates()

In [7]:
## we need to split transaction data into Dataframe/tabular structure as follow
new = bread['Date,Time,Transaction,Item'].str.split(',', n = 3, expand = True)

In [8]:
## assigning column to data frame "bread"
bread['Date'] = new[0]
bread['Time'] = new[1]
bread['Transaction'] = new[2]
bread['Item'] = new[3]

C:\Users\91892\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\91892\anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\91892\anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = v

In [9]:
# in this dataframe we only need column Trasaction and Item, rest is not needed in association mining rule
bread

Date,Time,Transaction,Item        Date      Time Transaction  \
0              2016-10-30,09:58:11,1,Bread  2016-10-30  09:58:11           1   
1       2016-10-30,10:05:34,2,Scandinavian  2016-10-30  10:05:34           2   
3      2016-10-30,10:07:57,3,Hot chocolate  2016-10-30  10:07:57           3   
4                2016-10-30,10:07:57,3,Jam  2016-10-30  10:07:57           3   
5            2016-10-30,10:07:57,3,Cookies  2016-10-30  10:07:57           3   
...                                    ...         ...       ...         ...   
21288      2017-04-09,14:32:58,9682,Coffee  2017-04-09  14:32:58        9682   
21289         2017-04-09,14:32:58,9682,Tea  2017-04-09  14:32:58        9682   
21290      2017-04-09,14:57:06,9683,Coffee  2017-04-09  14:57:06        9683   
21291      2017-04-09,14:57:06,9683,Pastry  2017-04-09  14:57:06        9683   
21292   2017-04-09,15:04:24,9684,Smoothies  2017-04-09  15:04:24        9684   

                Item  
0              Bread  
1       Scandinavian  
3      Hot chocolate  
4                Jam  
5            Cookies  
...              ...  
21288         Coffee  
21289            Tea  
21290         Coffee  
21291         Pastry  
21292      Smoothies  

[19640 rows x 5 columns]

In [10]:
# we need to convert cloumn transacton & item into Crosstab or we can say Binary Matrix as follow
tab = pd.crosstab(index= bread['Transaction'], columns= bread['Item'])
tab

Item         Adjustment  Afternoon with the baker  Alfajores  Argentina Night  \
Transaction                                                                     
1                     0                         0          0                0   
10                    0                         0          0                0   
100                   0                         0          0                0   
1000                  0                         0          0                0   
1001                  0                         0          0                0   
...                 ...                       ...        ...              ...   
995                   0                         0          0                0   
996                   0                         0          0                0   
997                   0                         0          0                0   
998                   0                         0          0                0   
999                   0                         0          1                0   

Item         Art Tray  Bacon  Baguette  Bakewell  Bare Popcorn  Basket  ...  \
Transaction                                                             ...   
1                   0      0         0         0             0       0  ...   
10                  0      0         0         0             0       0  ...   
100                 0      0         0         0             0       0  ...   
1000                0      0         0         0             0       0  ...   
1001                0      0         0         0             0       0  ...   
...               ...    ...       ...       ...           ...     ...  ...   
995                 0      0         0         0             0       0  ...   
996                 0      0         0         0             0       0  ...   
997                 0      0         0         0             0       0  ...   
998                 0      0         0         0             0       0  ...   
999                 0      0         0         0             0       0  ...   

Item         The BART  The Nomad  Tiffin  Toast  Truffles  Tshirt  \
Transaction                                                         
1                   0          0       0      0         0       0   
10                  0          0       0      0         0       0   
100                 0          0       0      0         0       0   
1000                0          0       0      0         0       0   
1001                0          0       0      0         1       0   
...               ...        ...     ...    ...       ...     ...   
995                 0          0       0      0         0       0   
996                 0          0       0      0         1       0   
997                 0          0       0      0         0       0   
998                 0          0       0      0         0       0   
999                 0          0       0      0         0       0   

Item         Valentine's card  Vegan Feast  Vegan mincepie  Victorian Sponge  
Transaction                                                                   
1                           0            0               0                 0  
10                          0            0               0                 0  
100                         0            0               0                 0  
1000                        0            0               0                 0  
1001                        0            0               0                 0  
...                       ...          ...             ...               ...  
995                         0            0               0                 0  
996                         0            0               0                 0  
997                         0            0               0                 0  
998                         0            0               0                 0  
999                         0            0               0              

In [11]:
## Just writing cdv file to check result
## I came to know that, we have one unwanted column named "NONE", we should remove it as follow and proceed further
#tab.to_csv('tab.csv')

In [12]:
## removing unwanted col "NONE"
tab = tab.drop(['NONE'], axis = 1)

In [13]:
tab

Item         Adjustment  Afternoon with the baker  Alfajores  Argentina Night  \
Transaction                                                                     
1                     0                         0          0                0   
10                    0                         0          0                0   
100                   0                         0          0                0   
1000                  0                         0          0                0   
1001                  0                         0          0                0   
...                 ...                       ...        ...              ...   
995                   0                         0          0                0   
996                   0                         0          0                0   
997                   0                         0          0                0   
998                   0                         0          0                0   
999                   0                         0          1                0   

Item         Art Tray  Bacon  Baguette  Bakewell  Bare Popcorn  Basket  ...  \
Transaction                                                             ...   
1                   0      0         0         0             0       0  ...   
10                  0      0         0         0             0       0  ...   
100                 0      0         0         0             0       0  ...   
1000                0      0         0         0             0       0  ...   
1001                0      0         0         0             0       0  ...   
...               ...    ...       ...       ...           ...     ...  ...   
995                 0      0         0         0             0       0  ...   
996                 0      0         0         0             0       0  ...   
997                 0      0         0         0             0       0  ...   
998                 0      0         0         0             0       0  ...   
999                 0      0         0         0             0       0  ...   

Item         The BART  The Nomad  Tiffin  Toast  Truffles  Tshirt  \
Transaction                                                         
1                   0          0       0      0         0       0   
10                  0          0       0      0         0       0   
100                 0          0       0      0         0       0   
1000                0          0       0      0         0       0   
1001                0          0       0      0         1       0   
...               ...        ...     ...    ...       ...     ...   
995                 0          0       0      0         0       0   
996                 0          0       0      0         1       0   
997                 0          0       0      0         0       0   
998                 0          0       0      0         0       0   
999                 0          0       0      0         0       0   

Item         Valentine's card  Vegan Feast  Vegan mincepie  Victorian Sponge  
Transaction                                                                   
1                           0            0               0                 0  
10                          0            0               0                 0  
100                         0            0               0                 0  
1000                        0            0               0                 0  
1001                        0            0               0                 0  
...                       ...          ...             ...               ...  
995                         0            0               0                 0  
996                         0            0               0                 0  
997                         0            0               0                 0  
998                         0            0               0                 0  
999                         0            0               0              

### Creating APRIORI_MY function to generate frequent itesets based on minimum threshold support = 0.02

In [14]:
def APRIORI_MY(data, min_support=0.05,  max_length = 5):
    import numpy as np
    import pandas as pd
    from itertools import combinations
    support = {}       
    L = list(data.columns)
    for i in range(1, max_length+1):
        c = list(combinations(L,i))
        L =[]
        for j in list(c):
            #print(j)
            sup = data.loc[:,j].product(axis=1).sum()/len(data.index)
            if sup > min_support:
                #print(sup, j)
                support[j] = sup
                L = list(set(L) | set(j))
    result = pd.DataFrame(list(support.items()), columns = ["Items", "Support"])
    return(result)

In [15]:
sup = APRIORI_MY(tab, 0.02, 2)
sup

Items   Support
0              (Alfajores,)  0.036093
1                  (Bread,)  0.324940
2                (Brownie,)  0.039765
3                   (Cake,)  0.103137
4                 (Coffee,)  0.475081
5                (Cookies,)  0.054034
6             (Farm House,)  0.038926
7          (Hot chocolate,)  0.057916
8                  (Juice,)  0.038296
9              (Medialuna,)  0.061379
10                (Muffin,)  0.038191
11                (Pastry,)  0.085510
12              (Sandwich,)  0.071346
13          (Scandinavian,)  0.028853
14                 (Scone,)  0.034309
15                  (Soup,)  0.034204
16                   (Tea,)  0.141643
17                 (Toast,)  0.033365
18              (Truffles,)  0.020145
19          (Toast, Coffee)  0.023502
20          (Juice, Coffee)  0.020460
21  (Hot chocolate, Coffee)  0.029378
22        (Cookies, Coffee)  0.028014
23            (Cake, Bread)  0.023187
24           (Cake, Coffee)  0.054349
25              (Cake, Tea)  0.023607
26          (Bread, Coffee)  0.089393
27             (Bread, Tea)  0.027909
28          (Bread, Pastry)  0.028958
29            (Coffee, Tea)  0.049523
30       (Coffee, Sandwich)  0.037981
31      (Coffee, Medialuna)  0.034939
32         (Coffee, Pastry)  0.047214

### Creating ASSOCIATION_RULE_MY function to generate itemset based on minimun threshold confidence.

In [17]:
def ASSOCIATION_RULE_MY(df, min_threshold=0.5):
    import pandas as pd
    from itertools import permutations
    #min_threshold = 0.5
    support = pd.Series(df.Support.values, index=df.Items).to_dict()
    data = []
    L= df.Items.values
    p = list(permutations(L, 2))

    for i in p:
        if set(i[0]).issubset(i[1]):
            conf = support[i[1]]/support[i[0]]
            #print(i, conf)
            if conf > min_threshold:
                #print(i, conf)
                j = i[1][not i[1].index(i[0][0])]
                lift = support[i[1]]/(support[i[0]]* support[(j,)])
                leverage = support[i[1]] - (support[i[0]]* support[(j,)])
                convection = (1 - support[(j,)])/(1- conf)
                data.append([i[0], (j,), support[i[0]], support[(j,)], support[i[1]], conf, lift, leverage, convection])

                        
    result = pd.DataFrame(data, columns = ["antecedents", "consequents", "antecedent support", "consequent support",
                                        "support", "confidence", "Lift", "Leverage", "Convection"])
    return(result)

In [20]:
my_asso = ASSOCIATION_RULE_MY(sup, 0.5)

### Finally sorting results by Lift to get highly associated itemsets.

In [22]:
my_asso.sort_values(by='Lift', ascending= False).head(10)

antecedents consequents  antecedent support  consequent support  \
7          (Toast,)   (Coffee,)            0.033365            0.475081   
4      (Medialuna,)   (Coffee,)            0.061379            0.475081   
5         (Pastry,)   (Coffee,)            0.085510            0.475081   
3          (Juice,)   (Coffee,)            0.038296            0.475081   
6       (Sandwich,)   (Coffee,)            0.071346            0.475081   
0           (Cake,)   (Coffee,)            0.103137            0.475081   
1        (Cookies,)   (Coffee,)            0.054034            0.475081   
2  (Hot chocolate,)   (Coffee,)            0.057916            0.475081   

    support  confidence      Lift  Leverage  Convection  
7  0.023502    0.704403  1.482699  0.007651    1.775789  
4  0.034939    0.569231  1.198175  0.005779    1.218561  
5  0.047214    0.552147  1.162216  0.006590    1.172079  
3  0.020460    0.534247  1.124537  0.002266    1.127031  
6  0.037981    0.532353  1.120551  0.004086    1.122468  
0  0.054349    0.526958  1.109196  0.005350    1.109667  
1  0.028014    0.518447  1.091280  0.002343    1.090053  
2  0.029378    0.507246  1.067704  0.001863    1.065276

## Cross Verifying results with  apriori and association rule from mlxtend -> frequent patterns package

In [25]:
from mlxtend.frequent_patterns import apriori, association_rules

In [27]:
## finding associate(item) which brought together with frequency greater than 0.1%
associate = apriori(df = tab, min_support= 0.02, use_colnames= True)

In [31]:
## Checking frequent itemsets
associate

support                 itemsets
0   0.036093              (Alfajores)
1   0.324940                  (Bread)
2   0.039765                (Brownie)
3   0.103137                   (Cake)
4   0.475081                 (Coffee)
5   0.054034                (Cookies)
6   0.038926             (Farm House)
7   0.057916          (Hot chocolate)
8   0.038296                  (Juice)
9   0.061379              (Medialuna)
10  0.038191                 (Muffin)
11  0.085510                 (Pastry)
12  0.071346               (Sandwich)
13  0.028853           (Scandinavian)
14  0.034309                  (Scone)
15  0.034204                   (Soup)
16  0.141643                    (Tea)
17  0.033365                  (Toast)
18  0.020145               (Truffles)
19  0.023187            (Cake, Bread)
20  0.089393          (Coffee, Bread)
21  0.028958          (Bread, Pastry)
22  0.027909             (Tea, Bread)
23  0.054349           (Coffee, Cake)
24  0.023607              (Cake, Tea)
25  0.028014        (Coffee, Cookies)
26  0.029378  (Coffee, Hot chocolate)
27  0.020460          (Juice, Coffee)
28  0.034939      (Coffee, Medialuna)
29  0.047214         (Coffee, Pastry)
30  0.037981       (Coffee, Sandwich)
31  0.049523            (Coffee, Tea)
32  0.023502          (Coffee, Toast)

### Createing associate rule such that item brought with conditional probability(Confidence) more than 50% with corresponding item

In [35]:
asso_rule = association_rules(associate, min_threshold= 0.5)

In [36]:
## Finally sorting results by Lift to get highly associated itemsets.

In [37]:
asso_rule.sort_values(by='lift', ascending= False).head(10)

antecedents consequents  antecedent support  consequent support  \
7          (Toast)    (Coffee)            0.033365            0.475081   
4      (Medialuna)    (Coffee)            0.061379            0.475081   
5         (Pastry)    (Coffee)            0.085510            0.475081   
3          (Juice)    (Coffee)            0.038296            0.475081   
6       (Sandwich)    (Coffee)            0.071346            0.475081   
0           (Cake)    (Coffee)            0.103137            0.475081   
1        (Cookies)    (Coffee)            0.054034            0.475081   
2  (Hot chocolate)    (Coffee)            0.057916            0.475081   

    support  confidence      lift  leverage  conviction  
7  0.023502    0.704403  1.482699  0.007651    1.775789  
4  0.034939    0.569231  1.198175  0.005779    1.218561  
5  0.047214    0.552147  1.162216  0.006590    1.172079  
3  0.020460    0.534247  1.124537  0.002266    1.127031  
6  0.037981    0.532353  1.120551  0.004086    1.122468  
0  0.054349    0.526958  1.109196  0.005350    1.109667  
1  0.028014    0.518447  1.091280  0.002343    1.090053  
2  0.029378    0.507246  1.067704  0.001863    1.065276

### Conclusion
 * Results from developed function(APRIORI_MY, ASSOCIATION_RULE_MY) has matched with builts packages.
 * it is observed that "Toast" & "Coffee" are highly associated with lift 1.48.
 * Coffee has been brought most frequently with 47.5% of all the transcaction

### Reference 
* "https://github.com/viktree/curly-octo-chainsaw/blob/master/BreadBasket_DMS.csv"
* https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwj-4qW15J_qAhUwzTgGHQ5MCuUQFjAHegQICRAB&url=http%3A%2F%2Fwww.ijarcs.info%2Findex.php%2FIjarcs%2Farticle%2Fdownload%2F4564%2F4083&usg=AOvVaw0tJaQUepruvpCogDKbi7T3